File: src/benchmark/framework/ScoringEngine.js
/**
 * Bayesian Beta-Binomial scoring engine for benchmark evaluation.
 *
 * Replaces passRate*N+M linear formula with:
 * - Beta posterior for proper uncertainty quantification
 * - Relative scoring against baseline/ceiling
 * - Consistency penalty for cross-scenario variance
 */

// ── Beta distribution utilities ────────────────────────────────────

/** Regularized incomplete beta function (simple numeric approx) */
function betaIncomplete(x, a, b) {
  // Use continued fraction approximation (Lentz's method)
  // For benchmark scoring, we need decent accuracy but not scientific precision
  if (x <= 0) return 0;
  if (x >= 1) return 1;
  // Simple numeric integration via trapezoidal rule (sufficient for our use case)
  const N = 200;
  const dx = x / N;
  let sum = 0;
  for (let i = 0; i <= N; i++) {
    const t = i * dx;
    const ft = Math.pow(t, a - 1) * Math.pow(1 - t, b - 1);
    sum += (i === 0 || i === N) ? ft / 2 : ft;
  }
  sum *= dx;
  // Normalize by Beta function B(a,b)
  const fullBeta = betaFn(a, b);
  return Math.min(1, Math.max(0, sum / fullBeta));
}

/** Beta function B(a,b) via log-gamma */
function betaFn(a, b) {
  return Math.exp(lnGamma(a) + lnGamma(b) - lnGamma(a + b));
}

/** Lanczos approximation to ln(Gamma(z)) */
function lnGamma(z) {
  const g = 7;
  const c = [0.99999999999980993, 676.5203681218851, -1259.1392167224028,
    771.32342877765313, -176.61502916214059, 12.507343278686905,
    -0.13857109526572012, 9.9843695780195716e-6, 1.5056327351493116e-7];
  if (z < 0.5) return Math.log(Math.PI / Math.sin(Math.PI * z)) - lnGamma(1 - z);
  z -= 1;
  let x = c[0];
  for (let i = 1; i < g + 2; i++) x += c[i] / (z + i);
  const t = z + g + 0.5;
  return 0.5 * Math.log(2 * Math.PI) + (z + 0.5) * Math.log(t) - t + Math.log(x);
}

/** Inverse Beta CDF via bisection (for credible intervals) */
function betaQuantile(p, a, b) {
  let lo = 0, hi = 1;
  for (let i = 0; i < 60; i++) {
    const mid = (lo + hi) / 2;
    if (betaIncomplete(mid, a, b) < p) lo = mid;
    else hi = mid;
  }
  return (lo + hi) / 2;
}

// ── Scoring functions ──────────────────────────────────────────────

/**
 * Compute Bayesian posterior statistics from a set of scores.
 * Uses Beta(2,2) weakly informative prior.
 *
 * @param {number[]} scores - Array of scores, each in [0,1]
 * @returns {{ mean, std, ci95: [number, number], median, p5, p95, posterior: {alpha, beta} }}
 */
export function bayesianScore(scores) {
  if (scores.length === 0) {
    return { mean: 0.5, std: 0.29, ci95: [0.05, 0.95], median: 0.5, p5: 0.05, p95: 0.95, posterior: { alpha: 2, beta: 2 } };
  }
  const sumX = scores.reduce((a, b) => a + b, 0);
  const alpha = 2 + sumX;
  const beta = 2 + scores.length - sumX;
  const mean = alpha / (alpha + beta);
  const variance = (alpha * beta) / ((alpha + beta) ** 2 * (alpha + beta + 1));
  const std = Math.sqrt(variance);
  return {
    mean: round(mean, 4),
    std: round(std, 4),
    ci95: [round(betaQuantile(0.025, alpha, beta), 4), round(betaQuantile(0.975, alpha, beta), 4)],
    median: round(betaQuantile(0.5, alpha, beta), 4),
    p5: round(betaQuantile(0.05, alpha, beta), 4),
    p95: round(betaQuantile(0.95, alpha, beta), 4),
    posterior: { alpha: round(alpha, 2), beta: round(beta, 2) },
  };
}

/**
 * Compute relative score: (agent - baseline) / (ceiling - baseline).
 * Returns 0 if agent <= baseline, 1 if agent >= ceiling.
 *
 * @param {number} agentScore
 * @param {number} baselineScore - Score from random/fallback strategy
 * @param {number} ceilingScore - Score from greedy oracle strategy
 * @returns {number} Relative score in [0,1]
 */
export function relativeScore(agentScore, baselineScore, ceilingScore) {
  const range = ceilingScore - baselineScore;
  if (range <= 0) return agentScore > baselineScore ? 1 : 0;
  return Math.max(0, Math.min(1, (agentScore - baselineScore) / range));
}

/**
 * Sandwich normalization (MeltingPot 2 / Agapiou et al. 2022) — array form.
 *
 *     score_norm(s) = (R_agent(s) − R_random(s)) / (R_exploiter(s) − R_random(s))
 *
 * R_random  ≡ deterministic-fallback policy (= Project-Utopia's `Guardrails` default).
 * R_exploiter ≡ scripted oracle policy (`ScriptedOraclePolicy.js`).
 *
 * Allows score > 1 to flag "superhuman LLM" runs without breaking math
 * (clip-on-output is the caller's choice). NaN-safe; range==0 short-circuits
 * to the binary above-baseline check.
 *
 * @param {number[]} agentScores
 * @param {number[]} fallbackScores  same length as agentScores (per-seed pairing)
 * @param {number[]} oracleScores    same length as agentScores
 * @param {{ clipUpperBound?: boolean }} [opts]  default: do NOT clip > 1 (paper convention)
 * @returns {number[]} normalized scores
 */
export function sandwichNormalize(agentScores, fallbackScores, oracleScores, opts = {}) {
  const n = agentScores.length;
  if (n === 0) return [];
  if (fallbackScores.length !== n || oracleScores.length !== n) {
    throw new Error(
      `sandwichNormalize: array length mismatch (agent=${n}, fb=${fallbackScores.length}, oracle=${oracleScores.length})`,
    );
  }
  const clipUpper = opts.clipUpperBound === true;
  const out = new Array(n);
  for (let i = 0; i < n; i++) {
    const range = oracleScores[i] - fallbackScores[i];
    let s;
    if (!Number.isFinite(range)) {
      // Non-finite oracle/fallback (NaN / Infinity) — return NaN to signal
      // upstream pipeline error rather than silently mask it.
      s = NaN;
    } else if (range < 0) {
      // Oracle is WORSE than fallback — this is an oracle blueprint bug.
      // Return NaN so caller can detect; do not silently produce a score.
      // (Reviewer B SC-1: collapsing this into "above-baseline binary" was wrong.)
      s = NaN;
    } else if (range === 0) {
      // Degenerate (oracle == fallback): binary above-baseline.
      s = agentScores[i] > fallbackScores[i] ? 1 : 0;
    } else {
      s = (agentScores[i] - fallbackScores[i]) / range;
      // Lower bound at 0 always; upper bound at 1 only if explicitly requested.
      if (s < 0) s = 0;
      if (clipUpper && s > 1) s = 1;
    }
    out[i] = Number.isFinite(s) ? round(s, 4) : s;
  }
  return out;
}

/**
 * Crafter geometric mean (Hafner 2021 — re-exported here for ScoringEngine consumers).
 * Punishes single-axis collapse.
 *
 *     S = exp((1/N) Σ ln(1 + sᵢ)) − 1
 */
export function geometricMean(scores) {
  if (!scores?.length) return 0;
  const N = scores.length;
  let logSum = 0;
  for (const s of scores) {
    const clipped = Math.max(0, Math.min(1, Number(s) || 0));
    logSum += Math.log(1 + clipped);
  }
  return round(Math.exp(logSum / N) - 1, 6);
}

/**
 * Apply consistency penalty: penalize high variance across scenarios.
 * finalScore = mean - lambda * std
 *
 * @param {number[]} scores
 * @param {number} [lambda=0.5]
 * @returns {number}
 */
export function consistencyAdjustedScore(scores, lambda = 0.5) {
  if (scores.length === 0) return 0;
  const mean = scores.reduce((a, b) => a + b, 0) / scores.length;
  const variance = scores.reduce((a, b) => a + (b - mean) ** 2, 0) / scores.length;
  return Math.max(0, mean - lambda * Math.sqrt(variance));
}

/**
 * Compute effect size (Cohen's d) between two groups.
 */
export function cohenD(groupA, groupB) {
  if (groupA.length === 0 || groupB.length === 0) return 0;
  const meanA = groupA.reduce((a, b) => a + b, 0) / groupA.length;
  const meanB = groupB.reduce((a, b) => a + b, 0) / groupB.length;
  const varA = groupA.reduce((a, b) => a + (b - meanA) ** 2, 0) / groupA.length;
  const varB = groupB.reduce((a, b) => a + (b - meanB) ** 2, 0) / groupB.length;
  const pooledStd = Math.sqrt((varA + varB) / 2);
  if (pooledStd === 0) return 0;
  return (meanA - meanB) / pooledStd;
}

/**
 * Compute Bayes Factor (BF10) approximation using BIC method.
 * Compares H1 (different means) vs H0 (same mean).
 */
export function bayesFactor(groupA, groupB) {
  const nA = groupA.length, nB = groupB.length;
  if (nA < 2 || nB < 2) return 1;
  const n = nA + nB;
  const all = [...groupA, ...groupB];
  const grandMean = all.reduce((a, b) => a + b, 0) / n;
  const sse0 = all.reduce((a, b) => a + (b - grandMean) ** 2, 0); // H0: single mean
  const meanA = groupA.reduce((a, b) => a + b, 0) / nA;
  const meanB = groupB.reduce((a, b) => a + b, 0) / nB;
  const sse1 = groupA.reduce((a, b) => a + (b - meanA) ** 2, 0)
             + groupB.reduce((a, b) => a + (b - meanB) ** 2, 0); // H1: two means
  // BIC approximation: BF10 ≈ exp((BIC0 - BIC1) / 2)
  const bic0 = n * Math.log(Math.max(1e-10, sse0 / n)) + 1 * Math.log(n); // 1 param
  const bic1 = n * Math.log(Math.max(1e-10, sse1 / n)) + 2 * Math.log(n); // 2 params
  return Math.exp((bic0 - bic1) / 2);
}

/**
 * Generate a comparison verdict from two groups of scores.
 *
 * @param {number[]} treatment - Scores from the treatment group
 * @param {number[]} control - Scores from the control group
 * @returns {{ deltaMean, cohenD, bayesFactor, verdict: string }}
 */
export function compareGroups(treatment, control) {
  const meanT = treatment.length ? treatment.reduce((a, b) => a + b, 0) / treatment.length : 0;
  const meanC = control.length ? control.reduce((a, b) => a + b, 0) / control.length : 0;
  const d = cohenD(treatment, control);
  const bf = bayesFactor(treatment, control);
  let verdict;
  if (bf > 10 && Math.abs(d) > 0.5) verdict = "CONFIRMED_IMPROVEMENT";
  else if (bf > 3 && Math.abs(d) > 0.3) verdict = "LIKELY_IMPROVEMENT";
  else if (bf < 1/3) verdict = "NO_EFFECT";
  else verdict = "AMBIGUOUS";
  return {
    deltaMean: round(meanT - meanC, 4),
    cohenD: round(d, 3),
    bayesFactor: round(bf, 2),
    verdict,
  };
}

function round(v, d = 2) {
  const s = Number(v);
  return Number.isFinite(s) ? Number(s.toFixed(d)) : s;
}

/**
 * HELM Mean Win Rate (Liang et al. 2022 / TMLR 2023, §F.4 of survey).
 *
 *   MWR_m = (1 / (|M| − 1)) · mean_d [ Σ_{m' ≠ m} 1{S_{m,d} > S_{m',d}} ] / |D|
 *         = (1 / (|D| · (|M| − 1))) · Σ_d Σ_{m' ≠ m} 1{S_{m,d} > S_{m',d}}
 *
 * In paper §4.4b we describe MWR as: "for each pair (m_i, m_j) compute the
 * fraction of dimensions where m_i beats m_j, then average across j ≠ i".
 * The two definitions are algebraically equivalent — we implement the
 * pairwise form because it cleanly handles per-pair missing dimensions
 * (skip the dim for that pair only) and exposes the per-pair win rate
 * for downstream non-transitivity / α-rank analysis.
 *
 * Tie handling: a strict `>` is used, matching HELM's convention. Ties
 * contribute 0 to the winner's count for that dim/pair (i.e., neither
 * model wins). Use a small ε in pre-processing if integer-equal ties
 * are expected.
 *
 * Missing dim handling: a dim is counted only for pairs where BOTH
 * agents have a finite score. The pair's denominator is the number of
 * jointly-present dims (so two agents that share zero dims contribute
 * NaN, which we surface as 0 for the MWR sum and a 0 pair contribution).
 *
 * @param {Object<string, Object<string, number>>} perAgentDimensionScores
 *   Shape: { [agentId]: { [dimKey]: score, … }, … }.
 *   All scores should already be normalized to a comparable benefit-form
 *   scale (use DimensionNormalizer first if not).
 * @returns {Object<string, number>} { [agentId]: mwr ∈ [0,1] }
 *   Single-agent input → returns { [agentId]: 0 } (no opponents).
 */
export function computeHelmMwr(perAgentDimensionScores) {
  const agentIds = Object.keys(perAgentDimensionScores ?? {});
  const out = {};
  if (agentIds.length === 0) return out;
  if (agentIds.length === 1) {
    // No opponents → MWR is undefined; report 0 for downstream pipeline safety.
    out[agentIds[0]] = 0;
    return out;
  }

  for (const i of agentIds) {
    let pairWinRateSum = 0;
    let pairCount = 0;
    const scoresI = perAgentDimensionScores[i] ?? {};
    for (const j of agentIds) {
      if (j === i) continue;
      const scoresJ = perAgentDimensionScores[j] ?? {};
      // Joint-dim set: dims where both agents have a finite score.
      const dims = new Set(Object.keys(scoresI));
      for (const d of Object.keys(scoresJ)) dims.add(d);
      let wins = 0;
      let jointDims = 0;
      for (const d of dims) {
        const a = Number(scoresI[d]);
        const b = Number(scoresJ[d]);
        if (!Number.isFinite(a) || !Number.isFinite(b)) continue;
        jointDims++;
        if (a > b) wins++;
      }
      if (jointDims === 0) {
        // No comparable dims → contribute 0 fraction (and the pair counts
        // toward the denominator, matching HELM's "always |M|−1 opponents"
        // convention; a dropped pair would inflate winners that simply
        // happen to share no dims with one opponent).
        pairWinRateSum += 0;
      } else {
        pairWinRateSum += wins / jointDims;
      }
      pairCount++;
    }
    out[i] = pairCount > 0 ? round(pairWinRateSum / pairCount, 4) : 0;
  }
  return out;
}

File: project_utopia/benchmark/framework/scoring_engine.py
"""Bayesian Beta-Binomial scoring engine (port of ``ScoringEngine.js``).

Replaces hand-rolled ``lnGamma`` / ``betaIncomplete`` / ``betaQuantile`` with
``scipy.stats.beta``. Numerical results agree with the JS implementation to
>= 4 decimal places on the standard fixtures (see ``test_bayesian_match.py``
and ``test_bayesian_score.py``).

The pure-Python helpers — :func:`sandwich_normalize`, :func:`geometric_mean`,
:func:`consistency_adjusted_score`, :func:`cohen_d`, :func:`bayes_factor`,
:func:`relative_score`, :func:`compare_groups`, :func:`compute_helm_mwr` — keep
the JS algorithms byte-for-byte. In particular :func:`compute_helm_mwr`
preserves the strict-``>`` tie rule, the per-pair joint-dim handling, and the
single-agent degenerate (0) return.
"""

from __future__ import annotations

import math
from typing import Any

from scipy import stats

__all__ = [
    "bayes_factor",
    "bayesian_score",
    "cohen_d",
    "compare_groups",
    "compute_helm_mwr",
    "consistency_adjusted_score",
    "geometric_mean",
    "relative_score",
    "sandwich_normalize",
]


# ---------------------------------------------------------------------------
# Numeric helpers
# ---------------------------------------------------------------------------


def _round(value: float, digits: int = 2) -> float:
    """Mirror the JS ``round(v, d)`` helper.

    NaN/Infinity pass through unchanged (matches JS ``Number.isFinite`` gate).
    """
    try:
        n = float(value)
    except (TypeError, ValueError):
        return value
    if not math.isfinite(n):
        return n
    return float(f"{n:.{digits}f}")


def _is_finite(x: Any) -> bool:
    try:
        return math.isfinite(float(x))
    except (TypeError, ValueError):
        return False


def _to_finite_float(x: Any) -> float:
    try:
        v = float(x)
    except (TypeError, ValueError):
        return float("nan")
    return v


# ---------------------------------------------------------------------------
# Bayesian score
# ---------------------------------------------------------------------------


def bayesian_score(scores: list[float]) -> dict[str, Any]:
    """Compute Beta(2,2)-prior posterior statistics from a [0,1] score vector.

    Returns the same JSON shape as the JS version: ``mean``, ``std``,
    ``ci95``, ``median``, ``p5``, ``p95``, ``posterior {alpha, beta}``.
    All numeric outputs are rounded to 4 decimal places (alpha/beta to 2)
    matching the JS rounding so direct numeric comparison is meaningful.

    Parameters
    ----------
    scores
        Iterable of floats, each ideally in ``[0,1]``.

    Returns
    -------
    dict
        Posterior summary plus ``posterior = {alpha, beta}``.

    Notes
    -----
    Empty input returns the prior (Beta(2,2)) summary statistics — the same
    defaults as JS.
    """
    if not scores:
        return {
            "mean": 0.5,
            "std": 0.29,
            "ci95": [0.05, 0.95],
            "median": 0.5,
            "p5": 0.05,
            "p95": 0.95,
            "posterior": {"alpha": 2, "beta": 2},
        }
    sum_x = float(sum(float(s) for s in scores))
    n = len(scores)
    alpha = 2.0 + sum_x
    beta = 2.0 + n - sum_x
    # scipy.stats.beta provides analytic mean/var/ppf.
    posterior = stats.beta(alpha, beta)
    mean = float(posterior.mean())
    std = float(posterior.std())
    return {
        "mean": _round(mean, 4),
        "std": _round(std, 4),
        "ci95": [
            _round(float(posterior.ppf(0.025)), 4),
            _round(float(posterior.ppf(0.975)), 4),
        ],
        "median": _round(float(posterior.ppf(0.5)), 4),
        "p5": _round(float(posterior.ppf(0.05)), 4),
        "p95": _round(float(posterior.ppf(0.95)), 4),
        "posterior": {"alpha": _round(alpha, 2), "beta": _round(beta, 2)},
    }


# ---------------------------------------------------------------------------
# Relative / sandwich scoring
# ---------------------------------------------------------------------------


def relative_score(agent_score: float, baseline_score: float, ceiling_score: float) -> float:
    """``(agent - baseline) / (ceiling - baseline)``, clipped to [0, 1].

    Degenerate range collapses to a binary above-baseline check, matching JS.
    """
    rng = float(ceiling_score) - float(baseline_score)
    if rng <= 0:
        return 1.0 if float(agent_score) > float(baseline_score) else 0.0
    val = (float(agent_score) - float(baseline_score)) / rng
    return max(0.0, min(1.0, val))


def sandwich_normalize(
    agent_scores: list[float],
    fallback_scores: list[float],
    oracle_scores: list[float],
    opts: dict[str, Any] | None = None,
) -> list[float]:
    """MeltingPot / Agapiou et al. 2022 sandwich normalization.

    ``S = (S_LLM - S_fallback) / (S_oracle - S_fallback)``

    The upper bound is **not** clipped by default (``S > 1`` flags a
    superhuman run); pass ``opts={'clip_upper_bound': True}`` to clip at 1.
    NaN-safe: degenerate range collapses to a binary above-baseline check,
    matching JS exactly. ``oracle < fallback`` returns NaN (oracle bug).
    """
    n = len(agent_scores)
    if n == 0:
        return []
    if len(fallback_scores) != n or len(oracle_scores) != n:
        raise ValueError(
            f"sandwich_normalize: array length mismatch "
            f"(agent={n}, fb={len(fallback_scores)}, oracle={len(oracle_scores)})"
        )
    clip_upper = bool(opts and opts.get("clip_upper_bound") is True)
    out: list[float] = []
    for i in range(n):
        rng = float(oracle_scores[i]) - float(fallback_scores[i])
        if not math.isfinite(rng):
            out.append(float("nan"))
            continue
        if rng < 0:
            # Oracle worse than fallback — oracle blueprint bug; surface NaN.
            out.append(float("nan"))
            continue
        if rng == 0:
            s = 1.0 if float(agent_scores[i]) > float(fallback_scores[i]) else 0.0
        else:
            s = (float(agent_scores[i]) - float(fallback_scores[i])) / rng
            if s < 0:
                s = 0.0
            if clip_upper and s > 1:
                s = 1.0
        out.append(_round(s, 4) if math.isfinite(s) else s)
    return out


# ---------------------------------------------------------------------------
# Aggregators
# ---------------------------------------------------------------------------


def geometric_mean(scores: list[float]) -> float:
    """Crafter geometric mean: ``exp(mean(log(1+s))) - 1``.

    Pure Python so it matches the JS computation byte-for-byte. Each ``s`` is
    clipped to ``[0, 1]`` first (defensive — same as JS).
    """
    if not scores:
        return 0.0
    n = len(scores)
    log_sum = 0.0
    for s in scores:
        try:
            v = float(s)
        except (TypeError, ValueError):
            v = 0.0
        if not math.isfinite(v):
            v = 0.0
        if v < 0:
            v = 0.0
        if v > 1:
            v = 1.0
        log_sum += math.log(1 + v)
    return _round(math.exp(log_sum / n) - 1, 6)


def consistency_adjusted_score(scores: list[float], lam: float = 0.5) -> float:
    """``max(0, mean - lambda * std)`` — same as JS."""
    if not scores:
        return 0.0
    mean = sum(float(s) for s in scores) / len(scores)
    variance = sum((float(s) - mean) ** 2 for s in scores) / len(scores)
    return max(0.0, mean - lam * math.sqrt(variance))


def cohen_d(group_a: list[float], group_b: list[float]) -> float:
    """Effect size Cohen's d. Pure Python."""
    if not group_a or not group_b:
        return 0.0
    mean_a = sum(float(s) for s in group_a) / len(group_a)
    mean_b = sum(float(s) for s in group_b) / len(group_b)
    var_a = sum((float(s) - mean_a) ** 2 for s in group_a) / len(group_a)
    var_b = sum((float(s) - mean_b) ** 2 for s in group_b) / len(group_b)
    pooled_std = math.sqrt((var_a + var_b) / 2)
    if pooled_std == 0:
        return 0.0
    return (mean_a - mean_b) / pooled_std


def bayes_factor(group_a: list[float], group_b: list[float]) -> float:
    """BF10 approximation via BIC. Pure Python."""
    n_a = len(group_a)
    n_b = len(group_b)
    if n_a < 2 or n_b < 2:
        return 1.0
    n = n_a + n_b
    all_vals = [float(x) for x in [*group_a, *group_b]]
    grand_mean = sum(all_vals) / n
    sse0 = sum((x - grand_mean) ** 2 for x in all_vals)
    mean_a = sum(float(s) for s in group_a) / n_a
    mean_b = sum(float(s) for s in group_b) / n_b
    sse1 = sum((float(s) - mean_a) ** 2 for s in group_a) + sum(
        (float(s) - mean_b) ** 2 for s in group_b
    )
    bic0 = n * math.log(max(1e-10, sse0 / n)) + 1 * math.log(n)
    bic1 = n * math.log(max(1e-10, sse1 / n)) + 2 * math.log(n)
    return math.exp((bic0 - bic1) / 2)


def compare_groups(treatment: list[float], control: list[float]) -> dict[str, Any]:
    """Generate a comparison verdict from two groups of scores (JS port)."""
    mean_t = sum(float(s) for s in treatment) / len(treatment) if treatment else 0.0
    mean_c = sum(float(s) for s in control) / len(control) if control else 0.0
    d = cohen_d(treatment, control)
    bf = bayes_factor(treatment, control)
    if bf > 10 and abs(d) > 0.5:
        verdict = "CONFIRMED_IMPROVEMENT"
    elif bf > 3 and abs(d) > 0.3:
        verdict = "LIKELY_IMPROVEMENT"
    elif bf < 1 / 3:
        verdict = "NO_EFFECT"
    else:
        verdict = "AMBIGUOUS"
    return {
        "deltaMean": _round(mean_t - mean_c, 4),
        "cohenD": _round(d, 3),
        "bayesFactor": _round(bf, 2),
        "verdict": verdict,
    }


# ---------------------------------------------------------------------------
# HELM Mean Win Rate (Layer 4b)
# ---------------------------------------------------------------------------


def compute_helm_mwr(
    per_agent_dimension_scores: dict[str, dict[str, float]] | None,
) -> dict[str, float]:
    """HELM Mean Win Rate (Liang et al. 2022 / TMLR 2023, §F.4).

    For each pair ``(m, m')`` where ``m != m'``:
      - ``joint_dims`` = dims where both agents have a finite score.
      - ``wins`` = count of joint dims where ``score(m, d) > score(m', d)``.
      - ``pairwise_rate`` = ``wins / joint_dims`` if joint_dims > 0 else 0.
    ``MWR(m)`` = mean of pairwise_rate over all opponents.

    Tie handling: strict ``>``; ties contribute 0 to neither side. Missing
    dims are joint-only. Single-agent input returns 0 for that agent. Empty /
    None input returns ``{}``.
    """
    if per_agent_dimension_scores is None:
        return {}
    scores_map = per_agent_dimension_scores
    if not isinstance(scores_map, dict):
        return {}
    agent_ids = list(scores_map.keys())
    out: dict[str, float] = {}
    if not agent_ids:
        return out
    if len(agent_ids) == 1:
        out[agent_ids[0]] = 0.0
        return out

    for i in agent_ids:
        scores_i = scores_map.get(i) or {}
        pair_winrate_sum = 0.0
        pair_count = 0
        for j in agent_ids:
            if j == i:
                continue
            scores_j = scores_map.get(j) or {}
            dims: set[str] = set(scores_i.keys()) | set(scores_j.keys())
            wins = 0
            joint = 0
            for d in dims:
                a = _to_finite_float(scores_i.get(d))
                b = _to_finite_float(scores_j.get(d))
                if not math.isfinite(a) or not math.isfinite(b):
                    continue
                joint += 1
                if a > b:
                    wins += 1
            if joint == 0:
                pair_winrate_sum += 0.0
            else:
                pair_winrate_sum += wins / joint
            pair_count += 1
        out[i] = _round(pair_winrate_sum / pair_count, 4) if pair_count > 0 else 0.0
    return out
